File: java/pdfclown.samples.cli/src/org/pdfclown/samples/cli/BookmarksParsingSample.java
package org.pdfclown.samples.cli;

import org.pdfclown.documents.Document;
import org.pdfclown.documents.Page;
import org.pdfclown.documents.fileSpecs.FileSpec;
import org.pdfclown.documents.interaction.actions.Action;
import org.pdfclown.documents.interaction.actions.GoToDestination;
import org.pdfclown.documents.interaction.actions.GoToEmbedded;
import org.pdfclown.documents.interaction.actions.GoToNonLocal;
import org.pdfclown.documents.interaction.actions.GoToURI;
import org.pdfclown.documents.interaction.actions.GoToEmbedded.TargetObject;
import org.pdfclown.documents.interaction.navigation.document.Bookmark;
import org.pdfclown.documents.interaction.navigation.document.Bookmarks;
import org.pdfclown.documents.interaction.navigation.document.Destination;
import org.pdfclown.files.File;
import org.pdfclown.objects.PdfObjectWrapper;

/**
  This sample demonstrates <b>how to inspect the bookmarks</b> of a PDF document.

  @author Stefano Chizzolini (http://www.stefanochizzolini.it)
  @version 0.1.0
*/
public class BookmarksParsingSample
  extends Sample
{
  @Override
  public boolean run(
    )
  {
    String filePath = promptPdfFileChoice("Please select a PDF file");

    // 1. Open the PDF file!
    File file;
    try
    {file = new File(filePath);}
    catch(Exception e)
    {throw new RuntimeException(filePath + " file access error.",e);}

    Document document = file.getDocument();

    // 2. Get the bookmarks collection!
    Bookmarks bookmarks = document.getBookmarks();
    if(bookmarks == null)
    {System.out.println("\nNo bookmark available (Outline dictionary not found).");}
    else
    {
      System.out.println("\nIterating through the bookmarks collection (please wait)...\n");
      // 3. Show the bookmarks!
      printBookmarks(bookmarks);
    }

    return true;
  }

  private void printBookmarks(
    Bookmarks bookmarks
    )
  {
    if(bookmarks == null)
      return;

    for(Bookmark bookmark : bookmarks)
    {
      // Show current bookmark!
      System.out.println("Bookmark '" + bookmark.getTitle() + "'");
      System.out.print("    Target: ");
      PdfObjectWrapper<?> target = bookmark.getTarget();
      if(target instanceof Destination)
      {printDestination((Destination)target);}
      else if(target instanceof Action)
      {printAction((Action)target);}
      else if(target == null)
      {System.out.println("[not available]");}
      else
      {System.out.println("[unknown type: " + target.getClass().getSimpleName() + "]");}

      // Show child bookmarks!
      printBookmarks(bookmark.getBookmarks());
    }
  }

  private void printAction(
    Action action
    )
  {
    System.out.println("Action [" + action.getClass().getSimpleName() + "] " + action.getBaseObject());
    if(action instanceof GoToDestination<?>)
    {
      if(action instanceof GoToNonLocal<?>)
      {
        FileSpec fileSpec = ((GoToNonLocal<?>)action).getFileSpec();
        if(fileSpec != null)
        {System.out.println("    Filename: " + fileSpec.getFilename());}

        if(action instanceof GoToEmbedded)
        {
          TargetObject target = ((GoToEmbedded)action).getTarget();
          System.out.println("    EmbeddedFilename: " + target.getEmbeddedFileName() + " Relation: " + target.getRelation());
        }
      }
      System.out.print("    ");
      printDestination(((GoToDestination<?>)action).getDestination());
    }
    else if(action instanceof GoToURI)
    {System.out.println("    URI: " + ((GoToURI)action).getURI());}
  }

  private void printDestination(
    Destination destination
    )
  {
    System.out.println(destination.getClass().getSimpleName() + " " + destination.getBaseObject());
    System.out.print("    Page ");
    Object pageRef = destination.getPageRef();
    if(pageRef instanceof Page)
    {
      Page refPage = (Page)pageRef;
      System.out.println((refPage.getIndex()+1) + " [ID: " + refPage.getBaseObject() + "]");
    }
    else
    {System.out.println(((Integer)pageRef+1));}
  }
}

File: dotNET/pdfclown.samples.cli/src/org/pdfclown/samples/cli/BookmarksParsingSample.cs
using org.pdfclown.documents;
using org.pdfclown.documents.fileSpecs;
using actions = org.pdfclown.documents.interaction.actions;
using org.pdfclown.documents.interaction.navigation.document;
using org.pdfclown.files;
using org.pdfclown.objects;

using System;

namespace org.pdfclown.samples.cli
{
  /**
    <summary>This sample demonstrates how to inspect the bookmarks of a PDF document.</summary>
  */
  public class BookmarksParsingSample
    : Sample
  {
    public override bool Run(
      )
    {
      // 1. Opening the PDF file...
      string filePath = PromptPdfFileChoice("Please select a PDF file");
      File file = new File(filePath);
      Document document = file.Document;

      // 2. Get the bookmarks collection!
      Bookmarks bookmarks = document.Bookmarks;
      if(bookmarks == null)
      {Console.WriteLine("\nNo bookmark available (Outline dictionary not found).");}
      else
      {
        Console.WriteLine("\nIterating through the bookmarks collection (please wait)...\n");
        // 3. Show the bookmarks!
        PrintBookmarks(bookmarks);
      }

      return true;
    }

    private void PrintBookmarks(
      Bookmarks bookmarks
      )
    {
      if(bookmarks == null)
        return;

      foreach(Bookmark bookmark in bookmarks)
      {
        // Show current bookmark!
        Console.WriteLine("Bookmark '" + bookmark.Title + "'");
        Console.Write("    Target: ");
        PdfObjectWrapper target = bookmark.Target;
        if(target is Destination)
        {PrintDestination((Destination)target);}
        else if(target is actions::Action)
        {PrintAction((actions::Action)target);}
        else if(target == null)
        {Console.WriteLine("[not available]");}
        else
        {Console.WriteLine("[unknown type: " + target.GetType().Name + "]");}

        // Show child bookmarks!
        PrintBookmarks(bookmark.Bookmarks);
      }
    }

    private void PrintAction(
      actions::Action action
      )
    {
      /*
        NOTE: Here we have to deal with reflection as a workaround
        to the lack of type covariance support in C# (so bad -- any better solution?).
      */
      Console.WriteLine("Action [" + action.GetType().Name + "] " + action.BaseObject);
      if(action.Is(typeof(actions::GoToDestination<>)))
      {
        if(action.Is(typeof(actions::GotoNonLocal<>)))
        {
          FileSpec fileSpec = (FileSpec)action.Get("FileSpec");
          if(fileSpec != null)
          {Console.WriteLine("    Filename: " + fileSpec.Filename);}

          if(action is actions::GoToEmbedded)
          {
            actions::GoToEmbedded.TargetObject target = ((actions::GoToEmbedded)action).Target;
            Console.WriteLine("    EmbeddedFilename: " + target.EmbeddedFileName + " Relation: " + target.Relation);
          }
        }
        Console.Write("    ");
        PrintDestination((Destination)action.Get("Destination"));
      }
      else if(action is actions::GoToURI)
      {Console.WriteLine("    URI: " + ((actions::GoToURI)action).URI);}
    }

    private void PrintDestination(
      Destination destination
      )
    {
      Console.WriteLine(destination.GetType().Name + " " + destination.BaseObject);
      Console.Write("    Page ");
      object pageRef = destination.PageRef;
      if(pageRef is Page)
      {
        Page refPage = (Page)pageRef;
        Console.WriteLine((refPage.Index+1) + " [ID: " + refPage.BaseObject + "]");
      }
      else
      {Console.WriteLine((int)pageRef+1);}
    }
  }
}